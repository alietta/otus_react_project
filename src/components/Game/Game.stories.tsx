import React from "react";
import { withKnobs, number, array, object } from "@storybook/addon-knobs";
import { Game } from "./Game";
import { DarkMode, LightMode } from "sancho";

export default {
  title: "Game",
  decorators: [withKnobs],
};

export const gameStory = () => (
  <div css={{ display: "flex" }}>
    <DarkMode>
      <Game />
    </DarkMode>
  </div>
);
