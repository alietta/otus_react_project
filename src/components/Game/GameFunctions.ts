export const preatyArray = (
  arr: Array<T>,
  length: number,
  filler: any
): Array<T> => {
  let copy = [...arr];
  if (copy.length > length) {
    copy = copy.slice(0, length);
  } else if (copy.length < length) {
    const addition = Array(length - copy.length).fill(filler);
    copy.splice(copy.length - 1, 0, ...addition);
  }
  return copy;
};

export const randomIntWithoutRepeat = (
  max: number,
  arr: Array<number>
): number => {
  const rand = Math.floor(Math.random() * Math.floor(max));
  if (arr.includes(rand)) {
    return randomIntWithoutRepeat(max, arr);
  }
  return rand;
};
export const makeCell = (
  cell: number,
  x: number,
  y: number
): { x: number; y: number } => {
  const line = Math.trunc(cell / x);
  return {
    y: line,
    x: cell - line * x,
  };
};
export const randomByPercent = (
  x: number,
  y: number,
  percent: number
): array => {
  const volume = x * y;
  const filled: Array<number> = [];
  const percentCount = Math.round((percent * volume) / 100);
  while (filled.length !== percentCount) {
    const cell = randomIntWithoutRepeat(volume, filled);
    filled.push(cell);
  }
  const filledArray = Array(y)
    .fill(false)
    .map(() => Array(x).fill(false));
  filled.forEach((cell) => {
    const pos = makeCell(cell, x, y);
    filledArray[pos.y][pos.x] = true;
  });
  return filledArray;
};
