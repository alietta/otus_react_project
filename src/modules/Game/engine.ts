export const checkNeighbor = (field: boolean[][], x: number, y: number) => {
  const rules = [
    [x - 1, y + 1],
    [x, y + 1],
    [x + 1, y + 1],
    [x - 1, y],
    [x + 1, y],
    [x - 1, y - 1],
    [x, y - 1],
    [x + 1, y - 1],
  ];
  const width = field[0].length;
  const height = field.length;
  const neighbor = rules.filter(([xRule, yRule]: [number, number]) => {
    const ruleInField =
      yRule >= 0 && xRule >= 0 && xRule < width && yRule < height;
    return ruleInField && field[yRule][xRule];
  });
  return [2, 3].includes(neighbor.length);
};

export const getGeneration = (field: boolean[][]): boolean[][] => {
  return field.map((row: boolean[], y: number) => {
    return row.map((life: boolean, x: number) => {
      return checkNeighbor(field, x, y);
    });
  });
};
