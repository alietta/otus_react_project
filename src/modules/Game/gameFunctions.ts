export const preatyArray = (
  arr: Array<T>,
  length: number,
  filler: T
): Array<T> => {
  let copy = [...arr];
  if (copy.length > length) {
    copy = copy.slice(0, length);
  } else if (copy.length < length) {
    const addition = Array(length - copy.length).fill(filler);
    copy.splice(copy.length - 1, 0, ...addition);
  }
  return copy;
};

export const randomIntWithoutRepeat = (
  max: number,
  arr: Array<number>
): number => {
  const rand = Math.floor(Math.random() * Math.floor(max));
  if (arr.includes(rand)) {
    return randomIntWithoutRepeat(max, arr);
  }
  return rand;
};
export const makeFieldPos = (data: {
  cell: number;
  width: number;
}): { x: number; y: number } => {
  const { cell, width } = data;
  const line = Math.trunc(cell / width);
  return {
    y: line,
    x: cell - line * width,
  };
};
export const randomByPercent = (
  x: number,
  y: number,
  percent: number
): number[] => {
  const volume = x * y;
  const filled: Array<number> = [];
  const percentCount = Math.round((percent * volume) / 100);
  while (filled.length !== percentCount) {
    const cell = randomIntWithoutRepeat(volume, filled);
    filled.push(cell);
  }
  return filled;
};

export const makeField = (
  x: number,
  y: number,
  filled: number[]
): boolean[][] => {
  const filledArray = Array(y)
    .fill(false)
    .map(() => Array(x).fill(false));
  filled.forEach((cell) => {
    const pos = makeFieldPos({ cell, width: x });
    filledArray[pos.y][pos.x] = true;
  });
  return filledArray;
};
