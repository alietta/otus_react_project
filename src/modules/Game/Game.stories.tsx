import React, { FC } from "react";
import { withKnobs } from "@storybook/addon-knobs";
import { Game } from "./Game";
import { DarkMode } from "sancho";

export default {
  title: "Game",
  decorators: [withKnobs],
};

export const gameStory: FC = () => (
  <div css={{ display: "flex" }}>
    <DarkMode>
      <Game />
    </DarkMode>
  </div>
);
