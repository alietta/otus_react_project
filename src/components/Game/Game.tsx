import React, { FunctionComponent, useState, useEffect } from "react";
import { useTheme, Layer } from "sancho";
import { GameMenu } from "../GameMenu";
import { GameField } from "../GameField";
import { preatyArray, randomByPercent } from "./GameFunctions";

const startMock = {
  field: {
    minWidth: 2,
    maxWidth: 5,
    minHeight: 2,
    maxHeight: 5,
  },
};
/* interface GameMenuProps { */
/* } */
/* const defaultProps: GameMenuProps = { */
/* }; */

const Game: FunctionComponent<> = () => {
  const theme = useTheme();
  const [fieldSize, setFieldSize] = useState<object>({
    width: startMock.minWidth,
    height: startMock.minHeight,
  });
  const [field, setField] = useState<Array>([[]]);
  const [percent, setPercent] = useState<number>(50);
  const [gameState, setGameState] = useState<object>({
    speed: 0,
    reset: false,
  });
  useEffect(() => {
    let newField = preatyArray(field, fieldSize.height, [false]);
    newField = newField.map((row) => preatyArray(row, fieldSize.width, false));
    setField(newField);
  }, [fieldSize]);

  useEffect(() => {
    if (gameState.reset === true) {
      const empty = Array(fieldSize.height)
        .fill(false)
        .map(() => Array(fieldSize.width).fill(false));
      setField(empty);
    }
  }, [gameState]);

  useEffect(() => {
    const newField = randomByPercent(
      startMock.field.minWidth,
      startMock.field.minHeight,
      50
    );
    setField(newField);
  }, []);

  const setFieldPercent = (perc: number): void => {
    const newField = randomByPercent(fieldSize.width, fieldSize.height, perc);
    setPercent(perc);
    setField(newField);
  };

  return (
    <Layer
      css={{
        overflow: "hidden",
        padding: theme.spaces.md,
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        width: "100%",
      }}
    >
      <div css={{ width: "100%", textAlign: "center" }}>
        <GameField field={field} setField={setField} />
      </div>
      <div css={{ minWidth: 422 }}>
        <GameMenu
          size={{ ...startMock.field, passSize: setFieldSize }}
          percent={{ onSubmit: setFieldPercent, startPercent: 50 }}
          controll={{ setGameState: setGameState }}
        />
      </div>
    </Layer>
  );
};

export { Game };
