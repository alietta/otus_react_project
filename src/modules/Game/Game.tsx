import React, { FunctionComponent, useState, useEffect } from "react";
import { useTheme, Layer } from "sancho";
import { GameMenu } from "components/GameMenu";
import { GameField } from "components/GameField";
import { preatyArray, randomByPercent } from "./gameFunctions";

const startMock = {
  field: {
    minWidth: 2,
    maxWidth: 20,
    minHeight: 2,
    maxHeight: 10,
  },
};
interface GameState {
  speed: number;
  reset: boolean;
}

const Game: FunctionComponent = () => {
  const theme = useTheme();
  const [fieldSize, setFieldSize] = useState<{ width: number; height: number }>(
    {
      width: startMock.minWidth,
      height: startMock.minHeight,
    }
  );
  const [field, setField] = useState<boolean[][]>([[]]);
  const [gameState, setGameState] = useState<GameState>({
    speed: 0,
    reset: false,
  });
  useEffect(() => {
    let newField = preatyArray(field, fieldSize.height, [false]);
    newField = newField.map((row) => preatyArray(row, fieldSize.width, false));
    setField(newField);
  }, [fieldSize]);

  useEffect(() => {
    if (gameState.reset === true) {
      const empty = Array(fieldSize.height)
        .fill(false)
        .map(() => Array(fieldSize.width).fill(false));
      setField(empty);
    }
  }, [gameState]);

  useEffect(() => {
    const newField = randomByPercent(
      startMock.field.minWidth,
      startMock.field.minHeight,
      50
    );
    setField(newField);
  }, []);

  const setFieldPercent = (perc: number): void => {
    const newField = randomByPercent(fieldSize.width, fieldSize.height, perc);
    setField(newField);
  };

  return (
    <Layer
      css={{
        overflow: "hidden",
        padding: theme.spaces.md,
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        width: "100%",
      }}
    >
      <div css={{ width: "100%", textAlign: "center" }}>
        <GameField field={field} setField={setField} />
      </div>
      <div css={{ minWidth: 422 }}>
        <GameMenu
          size={{ ...startMock.field, passSize: setFieldSize }}
          percent={{ onSubmit: setFieldPercent, startPercent: 50 }}
          controll={{ setGameState: setGameState }}
        />
      </div>
    </Layer>
  );
};

export { Game };
