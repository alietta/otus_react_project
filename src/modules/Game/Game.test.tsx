/**
 * @jest-environment jsdom
 */
import React from "react";
import { shallow, mount } from "enzyme";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";

import { Game } from "./Game";

const mockStore = configureMockStore([thunk]);

describe("Game", () => {
  let store;
  beforeEach(() => {
    store = mockStore({
      settings: {
        fieldSize: { width: 4, height: 4 },
        cellSize: { width: 10, height: 10 },
      },
      game: {
        status: "settings",
        speed: 0,
        filledCells: [],
      },
      field: [
        [false, false, false, false],
        [false, false, false, false],
        [false, false, false, false],
        [false, false, false, false],
      ],
    });
  });
  it("renders Game", () => {
    const elem = mount(<Game/>, {
      wrappingComponent: Provider,
      wrappingComponentProps: {
        store,
      },
    });
    expect(elem).toMatchSnapshot();
  });
});
