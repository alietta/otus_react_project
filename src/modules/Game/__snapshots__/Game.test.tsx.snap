// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders Game 1`] = `
<GameField
  field={
    Array [
      Array [],
    ]
  }
  setField={[Function]}
/>
<GameMenu
  controll={
    Object {
      "setGameState": [Function],
    }
  }
  percent={
    Object {
      "onSubmit": [Function],
      "startPercent": 50,
    }
  }
  size={
    Object {
      "maxHeight": 10,
      "maxWidth": 20,
      "minHeight": 2,
      "minWidth": 2,
      "passSize": [Function],
    }
  }
/>
`;
